repl/scala-2.10/src/main/scala/org/apache/spark/repl/SparkIMain.scala

```diff
@@ -978,15 +978,32 @@ import org.apache.spark.annotation.DeveloperApi
   @DeveloperApi
   def reset() {
     clearExecutionWrapper()
-    resetClassLoader()
     resetAllCreators()
+    resetToCheckpoint()
+  }
+  /**
+   * Save  previous Requests,referenced Names  and defined Names
+   * later these values can be used in reset operation
+   */
+    @DeveloperApi
+  def saveCheckpoint() {
+    prevRequestsCheckpoint      ++= prevRequests.clone()
+    referencedNameMapCheckpoint ++= referencedNameMap.clone()
+    definedNameMapCheckpoint    ++=definedNameMap.clone()
+  }
+  /**
+   * Reset prevRequests,referencedNameMap and definedNameMap to
+   * checkpointed state, so that it will only retain  spark initialization details
+   */
+  def resetToCheckpoint() {
     prevRequests.clear()
     referencedNameMap.clear()
     definedNameMap.clear()
-    virtualDirectory.delete()
-    virtualDirectory.create()
+    // Restore to checkpointed state
```

Comment:
Interesting, but this is removing some existing logic. Why should these be removed? there are also a number of style problems. I'm also under the impression that we try not to customize the shell more than is needed from Scala, so am a little skeptical of changing how it behaves here.